raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -2276,6 +2277,24 @@ public void resign(int epoch) {
         );
     }
 
+    private void validateSnapshotId(OffsetAndEpoch snapshotId) {
+        Optional<LogOffsetMetadata> highWatermarkOpt = quorum().highWatermark();
+        if (!highWatermarkOpt.isPresent() || highWatermarkOpt.get().offset < snapshotId.offset) {
+            throw new IllegalArgumentException("Trying to creating snapshot with invalid snapshotId: " + snapshotId + " whose offset is larger than the high-watermark: " +
```

Comment:
This might be better to split into two checks with different error messages. This would also let us avoid logging an `Optional`

Reply:
Fixed~